### PR TITLE
Allow admins to require a manual approval process before new accounts can be used (using MSC3866)

### LINKED ISSUE #13520: Implement MSC3866 (`M_USER_AWAITING_APPROVAL` error code)
Implement MSC3866 so that a homeserver can be configured to make it so new accounts must be approved before being usable.

An MVP of this implementation would consist in:

* a configuration flag to enable the feature (on top of the experimental flag)
* internal changes to keep track of an account's status wrt admin approval
* an admin API to allow admins to allow/deny accounts
* changes to the existing user APIs to allow checking if an account is approved, and listing all accounts pending approval

### PR DESCRIPTION
**Pull request by babolivier**
Resolves #13520 

Apologies about the size of this PR, but I wasn't sure it made sense to split it much more. Happy to discuss this if people have different opinions. It can be reviewed commit by commit though.

---

To enable this feature, add this to Synapse's configuration file:

```yaml
experimental_features:
  msc3866:
    enabled: true
    require_approval_for_new_accounts: true
```

Setting `require_approval_for_new_accounts` to `false` while keeping `enabled` to `true` allows manipulating a user's approval status without actively blocking unapproved users. `require_approval_for_new_accounts` is ignored if `enabled` is `false`.

If `require_approval_for_new_accounts` is `false`, all new users are considered already approved. This is to avoid users suddenly losing access to their accounts if the configuration changes to require approval for new users.

Users created by an admin (either by hitting the admin API directly or by using the `register_new_matrix_user` script) are automatically considered approved. However, it is possible for an admin to create an unapproved user by creating the account using the [Create or modify Account](https://matrix-org.github.io/synapse/latest/admin_api/user_admin_api.html#create-or-modify-account) admin API and set `approved` to `false` in the request's content (note that this parameter is ignored if `enabled` is `false`).

If the server is configured to require approval, a new user will be served a `403` error (with `ORG.MATRIX.MSC3866_USER_AWAITING_APPROVAL`) once the registration process finishes. If they try to log in before they have been approved, they will also be served a `403` error (also with `ORG.MATRIX.MSC3866_USER_AWAITING_APPROVAL`).

Admins can see a list of users pending approval by using the [List Accounts](https://matrix-org.github.io/synapse/latest/admin_api/user_admin_api.html#list-accounts) admin API with `approved=false` (this parameter is also ignored if `enabled` is `false`).

Admins can approve (or unapprove) a user by using the [Create or modify Account](https://matrix-org.github.io/synapse/latest/admin_api/user_admin_api.html#create-or-modify-account) admin API and set `approved` to `true` (for approved) or `false` (for unapproved) in the request's content (this parameter is also ignored if `enabled` is `false`).

---

In terms of future evolution for this feature, it would be nice to eg send emails to admins when there's a new user to approve and to figure out a way to contact users when they've been approved. This is out of the scope of this PR though.

### CODE REVIEW

**Review by DMRobertson — COMMENTED** [review]
My first instinct is that this sounds a little bit bespoke for Synapse. Isn't this something an identity provider would do if Synapse was using OIDC for its login and registration?

(I'll leave O to review)

**Review by reivilibre — CHANGES_REQUESTED** [review]
David has a point about this sounding like something that should be outside of Synapse's remit, either now or once OIDC-native is here, but I'm happy to believe that this has already been thought through somewhere.

As a note, I would have been tempted to leave the background job as a separate PR, since the rest of it can stand well on its own and that would have made it smaller to review.

I'm personally quite biased against background jobs (they're complicated, occasionally bugs pop up later down the line when someone upgrades very late and they can leave you with intermediate states which are harder to reason about); not saying it's wrong but on balance I'm not sure this one is worth it. It seems like we have to deal with the tri-state `NULL` value anyway because it's an intermediate state that the background job leaves you with, so perhaps rather omit the background job and leave `NULL` as meaning 'approved' implicitly.

File: synapse/storage/databases/main/registration.py
Comment (babolivier) [reply synapse/storage/databases/main/registration.py:1946]: Using `NULL` as approved is an interesting idea, though I'm a bit on the fence with this because imo it's very easy to read `NULL` on a boolean column as a false value, and it also means it needs a bit of special casing when handling it through the admin API (because then we need to explicitly make it `true`). I'd rather not use a default value since, as you point out, it involves rewriting the table in Postgres 10. That version reaching EOL in a couple of months is good news but I'd rather not block this PR on that.

It's also worth noting that this background update is basically a copy of the one we already have for setting the deactivation flag.

File: synapse/storage/databases/main/registration.py
Comment (babolivier) [reply synapse/storage/databases/main/registration.py:1994]: Gah, that was a leftover from my initial prototyping (since smallints are easier to deal with at that point).

File: synapse/rest/client/register.py
Comment (babolivier) [reply synapse/rest/client/register.py:769]: Oh that's a good point.

File: synapse/storage/databases/main/registration.py
Comment (babolivier) [reply synapse/storage/databases/main/registration.py:1949]: Actually since writing this PR a similar question came up in https://github.com/matrix-org/synapse/pull/13799, to which the decision was to treat `NULL` as `true`, which isn't _that_ much of a faff in the end. So let's go with that.

File: synapse/storage/databases/main/__init__.py
Comment (reivilibre) [inline synapse/storage/databases/main/__init__.py:206]: I find this arg name, both here in the method and in the admin API, non-obvious: it sounds like I ought to be able to write `approved=true` to get only approveds and `approved=false` to get only un-approveds.
I would be tempted to make it `Optional[bool]` so you get the choice of unfiltered or filtered either way.
Alternatively, to cater only to the 'get list of users needing approval' use case, I'd probably call it `needs_approval` and invert the meaning.

File: synapse/rest/admin/users.py
Comment (reivilibre) [inline synapse/rest/admin/users.py:105]: mentioned elsewhere but would suggest `Optional[bool]` for this interface so admin frontends can filter any way they like

File: synapse/rest/admin/users.py
Comment (reivilibre) [inline synapse/rest/admin/users.py:395]:
```suggestion
            new_user_approved = self._msc3866_enabled and approved is True
```

Comment (reivilibre) [reply synapse/rest/admin/users.py:395]: Worth noting that we are consciously requiring even admins to specify `approved = True` rather than assuming that an admin creating the account at all is sufficient for approval by default.

Should this either:
- just approve the user unless overridden
- not depend on whether `require_approval_for_new_accounts` is True?

Thinking about it: if we unlock this MSC as-is, the current implementation will break existing admin API uses.

File: synapse/storage/databases/main/registration.py
Comment (reivilibre) [inline synapse/storage/databases/main/registration.py:185]: (!) could be None

File: tests/storage/test_registration.py
Comment (reivilibre) [inline tests/storage/test_registration.py:214]: since `get_user_by_id` is deprecated, please could you update this to use `get_userinfo_by_id` instead?

At the same time, you can do the `None → True` conversion whilst loading it into the `UserInfo` struct.

File: tests/storage/test_registration.py
Comment (reivilibre) [inline tests/storage/test_registration.py:216]: this smells like it wouldn't be portable to Postgres?

File: synapse/storage/schema/main/delta/73/03users_approved_column.sql
Comment (reivilibre) [inline synapse/storage/schema/main/delta/73/03users_approved_column.sql:17]: Please describe what `NULL` means here, within the schema delta — it's our best place to look for definitions of columns.

**Review by reivilibre — CHANGES_REQUESTED** [review]
I know I've left a flurry of comments, but this is basically good at this point.

Really minor nitty stuff. Feel free to ping when you want your re-review, sorry for missing this :(

File: tests/storage/test_registration.py
Comment (reivilibre) [reply tests/storage/test_registration.py:216]: (but CI is green, so not sure why this passes actually.. When you use `UserInfo` you can just `assertTrue` though, which feels a bit better of an interface to expose)

File: synapse/config/experimental.py
Comment (babolivier) [reply synapse/config/experimental.py:24]: Looks like we're only doing it in http handlers. Also I'm not sure I see the point of validation to this extent for experimental, if-you-use-this-without-knowing-exactly-what-you're-doing-you're-on-your-own options.

File: synapse/storage/databases/main/__init__.py
Comment (babolivier) [reply synapse/storage/databases/main/__init__.py:206]: Good point, though the reason I'm doing it like this is to be consistent with other parameters such as `guests`. I think it's clearer that way because similar parameters imply that if they're absent from the request then the category of users they refer to will be included in the response (whereas `needs_approval` kind of implies otherwise). But then we also have the `deactivated` parameter which works kind of the same way but with flipped values of true/false and aaargh why is life so terrible.
But in general I still think it makes more sense to have an `approved` parameter.

File: synapse/config/experimental.py
Comment (DMRobertson) [reply synapse/config/experimental.py:24]: We're free to use it wherever we want to. The original suggestion was to use it for config and I even prototyped this: https://github.com/matrix-org/synapse/issues/12651#issuecomment-1133940984

We don't have many examples of using Pydantic yet so I wouldn't force it on anyone. We would probably need some glue code for config too?

But I think it is pretty good at what it does, gives better error messages earlier and gives us an opportunity to propagate more types through Synapse.

File: synapse/rest/admin/users.py
Comment (babolivier) [reply synapse/rest/admin/users.py:395]: > Worth noting that we are consciously requiring even admins to specify `approved = True` rather than assuming that an admin creating the account at all is sufficient for approval by default.

Thanks for pointing that out, that's a thinko from me. Users created by the admin API should be considered approved. Plus the condition for `new_user_approved` meant that if the feature isn't enabled new users are considered not approved, which is incorrect.

Comment (babolivier) [reply synapse/rest/admin/users.py:395]: So

> just approve the user unless overridden

Is the correct behaviour.

File: tests/storage/test_registration.py
Comment (babolivier) [reply tests/storage/test_registration.py:216]: Turns out `1 == True` in Python :neutral_face:

Comment (DMRobertson) [reply tests/storage/test_registration.py:216]: I think you want `assertIs(expr, True)`, see 

https://docs.python.org/3/library/unittest.html?highlight=asserttrue#unittest.TestCase.assertTrue

Comment (DMRobertson) [reply tests/storage/test_registration.py:216]: That is: `assertTrue` ought to be called `assertTruthy`

File: tests/storage/test_registration.py
Comment (babolivier) [reply tests/storage/test_registration.py:214]: I'd rather test the function that's being modified directly, rather than doing this indirectly by testing a function that calls it. Plus there are still loads of places calling `get_user_by_id` and not `get_userinfo_by_id` in the code.

File: tests/storage/test_registration.py
Comment (reivilibre) [reply tests/storage/test_registration.py]: > Turns out `1 == True` in Python neutral_face

Wait what? I don't know how I didn't know that. I know JS is like that but I for some reason thought Python didn't do it that way.... Odd..

File: tests/storage/test_registration.py
Comment (reivilibre) [reply tests/storage/test_registration.py:214]: just feels like we're propagating the use of a deprecated function here rather than its light wrapper which is the nominated successor. But OK, all the more reason to remove it soon

File: synapse/storage/databases/main/__init__.py
Comment (reivilibre) [reply synapse/storage/databases/main/__init__.py:206]: I see the dilemma here for either doing it 'consistently silly' or being inconsistent but maybe more useful.

I think I'd still prefer to do it inconsistently but less sillily, but don't feel too strongly about it so if you're not convinced that's fine by me.

File: synapse/config/experimental.py
Comment (reivilibre) [reply synapse/config/experimental.py:24]: No strong opinion but presumably this won't always be an experimental option, at which point it will just get renamed into something unexperimental? So I don't really buy that we should do things less properly for experimental options, but it's only a minor thought

File: tests/rest/admin/test_user.py
Comment (reivilibre) [inline tests/rest/admin/test_user.py:2659]: Given that these are misnomers, I would like to second dmr's suggestion of `assertIs(.., False)` and similar for `True` to make doubly sure we're not giving 0s and 1s back. Big :/ that these test helpers are badly named.

**Review by reivilibre — APPROVED** [review]

File: synapse/storage/databases/main/__init__.py
Comment (babolivier) [reply synapse/storage/databases/main/__init__.py:206]: I do agree that it would be worth changing this API to fix this because it's confusing, but I'd rather we do it once in a consistent way (in a separate PR) rather than introduce inconsistencies that could confuse the users even more.

File: tests/rest/admin/test_user.py
Comment (babolivier) [reply tests/rest/admin/test_user.py:2659]: Sure. Though this means fixing the API to return boolean values for `admin`, `shadow_banned`, `deactivated` and `approved`, which it wouldn't do on its own (but happy to do that).

**Comment by babolivier** [comment]
Re-requesting a review since I've implemented a recent change in the MSC (https://github.com/matrix-org/matrix-spec-proposals/pull/3866/commits/3df48d3349538081cc015948a0be12ccb783707a, with https://github.com/matrix-org/matrix-spec-proposals/pull/3866/commits/51fd1ed6e57566e138aa9099cd2e4f6e77d3d711 adding missing unstable IDs). Should be fairly straightforward.

File: synapse/storage/databases/main/registration.py
Comment (reivilibre) [inline synapse/storage/databases/main/registration.py:206]: slightly worrying as `bool(None) is False`. I think this is OK now because it's all coalesced, but seems like an easy way to trip up in future. Probably OK as I guess it will be applied consistently and a mistake will be noticed early on..?

**Review by reivilibre — APPROVED** [review]

File: synapse/storage/databases/main/registration.py
Comment (babolivier) [reply synapse/storage/databases/main/registration.py:206]: Yeah good point, adding a comment to try to catch this in the future.

**Pull request merged**
